ipfs_api_mount/__init__.py

```diff
@@ -54,7 +53,6 @@ def resolve_path(object_path):
                 raise InvalidIPFSPathException()
             return absolute_path[6:]
 
-        @lru_cache(maxsize=object_data_cache_size)
```

Comment:
object_data should be cached, I think. Without cache reading from large files causes 2-3 times more calls. Files are composed from many objects organized in a tree. To read data from a leaf you have to traverse all nodes above. (This is theory. I don't have any measurements.)

Reply:
object_data doesn't need to be cached. The IPFS-Node does the local caching. If we cache this, we're basically copying data we request to another location in memory and it's never used again since the cache is LRU and will be discarded after some other read requests.

Add a debug message for a cache hit, and let me know if there are any, I strongly doubt that.

Reply:
IPFS node does the caching - agreed - but the same is true for any other kind of data we read from IPFS node. Purpose of the cache is not to minimize traffic between IPFS node and IPFS network, but to minimize traffic between IPFS node and fuse handling process. Thats how I understand it.

I'll check the cache hits as you said and will let you know.

cheers

Reply:
Accessing data via a CID is pretty lightweight for the IPFS node, accessing a full directory listing, on the other hand, data has to be parsed and a lot of reads have to be done, to extract the information you request from a node.

When you got a folder with 5000 files above 256 KByte in size, the IPFS-daemon has to read 1.22 GB from the disk for a single directory listing.

I'm not sure that the IPFS-client do cache a directory listing, so it depends purely on caching of the 1.22 GB by the kernel to speedup the next listing - while the processing of the GB is still needed.

That's why we should cache the endresults of all meta-data requests (except very basic ones).

Reply:
There is already a test for measuring how many requests are made when reading a file: https://github.com/SupraSummus/ipfs-api-mount/blob/master/tests/test_performance.py#L10
This test won't pass with object link cache disabled. The settings for this test are a bit pessimistic (small IPFS chunks, small reads) - but this depicts what for is the object link cache.

I'm strongly against removing link cache. Instead we can give an option to disable it. Users can choose what they like.

With current code, if you don't want link caching, you can pass `--link-cache-size 1`. Zero will throw an error.

Reply:
I haven't digged into the test, but using 4K seems to be extremely unrealistic. There's a massive overhead doing something like this.

For your test you should stick to the default block size of 256K and read like 10 GB from the IPFS instead of just 400 K.

Test out what's faster, with or without a cache. :)

Reply:
Ah

```
self.assertGreaterEqual(measuring_client.request_count, chunk_count)
self.assertLess(measuring_client.request_count, 1.2 * chunk_count)
```

Without a cache, this is obviously no longer deterministic.

Reply:
Yep, this test setting is unrealistic. It is such to make automated tests run fast. Real files easily can be 64 times larger than test files and then 256k block size will have the same effects as 4k test blocks.

Also, I wouldn't assume real-world blocks will be always 256k. Files can be chunked by custom chunkers, can be patched after creation and so on and so on. Realy I think 256k are almost a best-case scenario.

> Without a cache, this is obviously no longer deterministic.

What do you mean by that?

Reply:
I did some testing.

Test settings:
 * commit 5353e706b898769222583d4edda064d8983b80de
 * go-ipfs version: 0.4.23-6ce9a355f (running on localhost)
 * 100MiB of random data
 * mount options: `--no-threads --background`

Results:
 * without any changes to the code: real 0m1.754s
 * with `subblock_cids_cache` and `subblock_sizes_cache` disabled (no hit checking and no writing to the cache): real 0m11.272s
 * with `block_cache` disabled (no hit checking and no writing to the cache): real 0m11.930s

So, my conclusion is that link cache and block data cache do have impact on performance.